pyroute2/netlink/rtnl/rtmsg.py

```diff
@@ -160,7 +160,90 @@ class seg6_encap_info(nla):
+
+            __slots__ = ()
+
+            fields = (('encapmode', 'I'),
+                      ('nexthdr', 'B'),
+                      ('hdrlen', 'B'),
+                      ('type', 'B'),
+                      ('segments_left', 'B'),
+                      ('first_segment', 'B'),
+                      ('flags', 'B'),
+                      ('reserved', 'H'),
+                      ('segs', 's'))
+
+            # Corresponding values for seg6 encap modes
+            SEG6_IPTUN_MODE_INLINE = 0
+            SEG6_IPTUN_MODE_ENCAP = 1
+
+            # Mapping string to nla value
+            encapmodes = {
+              "inline": SEG6_IPTUN_MODE_INLINE,
+              "encap": SEG6_IPTUN_MODE_ENCAP
+            }
+
+            # nla value for seg6 type
+            SEG6_TYPE = 4
+
+            def get_family(self):
+                pointer = self
+                while pointer.parent is not None:
+                    pointer = pointer.parent
+                return pointer.get('family', AF_UNSPEC)
+
+            def encode(self):
+                # Retrieve the family
+                family = self.get_family()
+                # Seg6 can be applied only to IPv6
+                if family == AF_INET6:
+                    # Get mode
+                    mode = self['mode']
+                    # Get segs
+                    segs = self['segs']
+                    # Add mode to value
+                    self['encapmode'] = self.encapmodes.get(mode, self.SEG6_IPTUN_MODE_ENCAP)
+                    # Calculate srlen
+                    srhlen = 8 + 16 * len(segs);
+                    # Calculate and set hdrlen
+                    self['hdrlen'] = (srhlen >> 3) - 1
+                    # Add seg6 type
+                    self['type'] = self.SEG6_TYPE
+                    # Add segments left
+                    self['segments_left'] = len(segs) - 1
+                    # Add fitst segment
+                    self['first_segment'] = len(segs) - 1
+                    # Init segs_value
+                    segs_value = b''
+                    # Iterate over segments
+                    for seg in segs:
+                      # Convert to network byte order and add to value
+                      segs_value += inet_pton(family, seg)
+                    # Set the converted string
+                    self['segs'] = segs_value
+                else:
+                    raise TypeError('Family %s not supported for seg6 tunnel'
+                                    % family)
+                nla.encode(self)
+
+            def decode(self):
+                # Decode the data
+                nla.decode(self)
+                # Calculate offset of the segs
+                offset = self.offset + 16
+                # Point the addresses
+                addresses = self.data[offset:]
+                # Init segs
+                segs = []
+                # Move 128 bit in each step
+                for i in range(len(addresses) // 16):
```

Comment:
We should extract here `segments_left + 1`, since `self.data`  contains all the message, not only this NLA. Something like:
```python
                 for i in range(len(addresses) // 16):                            
                     if i > self['segments_left']:                                
                         break                                                    
                     # Save the segment                                           
                     segs.append(inet_ntop(AF_INET6, addresses[i*16:i*16+16]))    
```

Reply:
Yes, I was planning to do that in order to proper support HMAC functionality which introduces other data after segments.

https://git.kernel.org/pub/scm/linux/kernel/git/shemminger/iproute2.git/tree/ip/iproute_lwtunnel.c#n435

https://git.kernel.org/pub/scm/linux/kernel/git/shemminger/iproute2.git/tree/ip/iproute_lwtunnel.c#n455